src/frontmenu_ingame_tabs.c

```diff
@@ -1769,7 +1769,9 @@ void maintain_event_button(struct GuiButton *gbtn)
     if ((dungeon->visible_event_idx != 0) && (evidx == dungeon->visible_event_idx))
     {
         turn_on_event_info_panel_if_necessary(dungeon->visible_event_idx);
-        if (is_game_key_pressed(Gkey_ToggleMessage, &keycode, false))
+        //TODO: that should be not here, Keys should be processed at one place
```

Comment:
Where do you suggest I put this?

Reply:
Somewhere near "keys".
This is actually a "button update handler" - it is called from ui drawing code.
In this file should be a callback function called from input-related file or a polling function called from there.

Reply:
Sorry, you might as well be speaking russian here.

Reply:
I mean we should group all keyboard input towards one layer and one place in codebase.
 This function `maintain_event_button` is only used in GuiButtonInit structure as `maintain callback`. It is a drawing abstraction "lets check how should we draw a specific button". 
But here we implement some input logic instead. 
That is not good.

Reply:
I'm new at all this, so while I understand what you're saying, I don't know how to fix this. I'm open to suggestions.

Reply:
Actually we have to make some kind of refactoring. 
Good way to do this is to have 'keyboard layout' abstraction.
I.e. when you are in first person mode you have one layout, when you are in dungeon mode you have another etc.
So it should be two mappings `keyboard shortcut -> binding name` and `binding name -> function`

Then it is good to have a function like`void register_key_event(struct KeyboardLayout *layout, enum BindingName binding, void (*) (void *context), void *context);`
Each "component" should have some "init" function where it should register its events.
So when you have some transition in game state (i.e. player possess a creature or open a menu) Game should just set active layout.
And main loop should just call something like `process_keyboard_event(active_layout, keycode, modifiers);`